### PR TITLE
libsyntax_ext: Lint on multiple test & bench attrs

### LINKED ISSUE #62847: test attribute doesn't check if another test attribute is specified
Currently if you slap `#[test]` two times you're going to get double tests in `cargo test` output.
I suspect it doesn't really care about presence of other attributes.
And I only noticed it because `tokio::test` attribute that wraps async functions for  tests also appends own `#[test]`

Not sure if it is bug and diagnostic message should be presented to user?

### PR DESCRIPTION
**Pull request by DoumanAsh**
Added errors on multiple `test` and `bench` attrs.

Not sure if there are test for these attributes though?

Fixes #62847

### CODE REVIEW

**Comment by nikomatsakis** [comment]
I'm a bit wary of breaking existing code like this. We might want to issue a warning instead. I'm going to nominate briefly for @rust-lang/lang to discuss. In the meantime, perhaps a crater run will shed light.

@bors try

**Comment by DoumanAsh** [comment]
I'm not sure that this was actually intended usage of test/bench attribute in the first place.
So I don't think that if someone used this bug, it is a problem for us to fix it.

I didn't look deeply but it produces quite unexpected output in `cargo test`, but I didn't check whether tests run twice or not.

**Comment by nikomatsakis** [comment]
> I'm not sure that this was actually intended usage of test/bench attribute in the first place.

It's certainly not the intended usage, but if people are out in the wild relying on it, we try not to create errors without a good reason.

**Comment by joshtriplett** [comment]
Discussed in the lang team meeting. Summary:

We think the current behavior makes sense, though it's unlikely to be what people want. You'd get similar behavior if you write a proc macro twice.

We'd be all for making this a lint. Such a lint could potentially warn about specific other multiply-used attributes, such as two `must_use` or `should_panic` attributes.

**Comment by DoumanAsh** [comment]
>You'd get similar behavior if you write a proc macro twice.

This is implementation defined behavior, because proc macro might throw error on invalid usage of multiple attributes.

If team decided for lint, should this PR be closed? Not sure if clippy is separate repo or not.

**Comment by Centril** [comment]
> If team decided for lint, should this PR be closed? Not sure if clippy is separate repo or not.

Up to you; if you prefer you can change this PR into using the lint machinery in the compiler or you can file another PR in this repo.

**Comment by DoumanAsh** [comment]
Ok, but by lints do you mean compiler warning level or clippy lints? I'm not really familiar with Rust internals

**Comment by Centril** [comment]
@DoumanAsh A compiler lint, e.g. like those in https://github.com/rust-lang/rust/blob/master/src/librustc_lint/.

**Comment by DoumanAsh** [comment]
@Centril I used buffered lint as it would seem what is used for when macro expansion happens.
Let me know if I got it wrong.
As for lint I choosen Ill formed attribute as it seems to be fitting case.
After all it is not valid to use multiple test/bench attributes.

**Comment by DoumanAsh** [comment]
Also had to rebase

**Comment by edmilsonefs** [comment]
Hey! This is a ping from triage, we would like to know if you @petrochenkov could give us a few minutes to share your thoughts on it.

**Comment by petrochenkov** [comment]
Low priority, deferred until the weekend (unless I deal with other reviews earlier).

I don't like this in general, probably not useful enough for a built-in lint.

**Comment by DoumanAsh** [comment]
>I don't like this in general, probably not useful enough for a built-in lint.

Since this is not intended usage of these attributes compiler should error/warn about it

**Comment by petrochenkov** [comment]
@DoumanAsh 
So the mistake that happened in practice (https://github.com/rust-lang/rust/issues/62847) looked like this?
```rust
#[test]
#[tokio::test]
fn foo() {}
```
The PR as implemented won't catch that, it'll catch only something written literally like
```rust
#[test]
#[test]
fn foo() {}
```

File: src/libsyntax_ext/test.rs [hunk 1]
```diff
@@ -234,8 +235,19 @@ fn should_panic(cx: &ExtCtxt<'_>, i: &ast::Item) -> ShouldPanic {
...
     let ref sd = cx.parse_sess.span_diagnostic;
+
+    if attr::contains_name(&i.attrs, sym::test) {
+        cx.parse_sess.buffer_lint(
+            BufferedEarlyLintId::IllFormedAttributeInput,
+            i.span,
+            i.id,
+            "multiple test attributes, only one is needed"
+        );
+        return false;
```
Comment (petrochenkov) [inline src/libsyntax_ext/test.rs]: We are reporting a warning, not an error, and this is technically legal just suspicious, so we shouldn't return `false` here, but should rather report and continue.
(Also, if `false` is not returned this no longer belongs to `fn has_test_signature` logically.)

File: src/libsyntax_ext/test.rs [hunk 2]
```diff
@@ -287,6 +299,18 @@ fn has_test_signature(cx: &ExtCtxt<'_>, i: &ast::Item) -> bool {
...
+    let ref sd = cx.parse_sess.span_diagnostic;
+
+    if attr::contains_name(&i.attrs, sym::bench) {
+        cx.parse_sess.buffer_lint(
+            BufferedEarlyLintId::IllFormedAttributeInput,
+            i.span,
+            i.id,
+            "multiple bench attributes, only one is needed"
+        );
+        return false;
```
Comment (petrochenkov) [inline src/libsyntax_ext/test.rs]: Same comment as https://github.com/rust-lang/rust/pull/62850/files#r310388408.

**Comment by petrochenkov** [comment]
Could you add a test with two test cases - `#[test] #[test] fn foo() {}` and `#[rustfmt::test] #[test] fn foo() {}`?

File: src/libsyntax_ext/test.rs (hunk 1 above)
Comment (DoumanAsh) [reply src/libsyntax_ext/test.rs]: ok

**Comment by DoumanAsh** [comment]
@petrochenkov `tokio::test` adds `#[test]` so it would catch

Could you elaborate where are test cases for macro attributes?

**Comment by petrochenkov** [comment]
`src/test/ui/attributes`, for example.

**Comment by petrochenkov** [comment]
If this doesn't introduce a new lint, but rather reuses `ill_formed_attribute_input`, then this is ok, I guess.
The warning doesn't introduce too much technical debt and can be removed if stands in the way of any future work (e.g. converting built-in macros into regular proc macros).

**Comment by petrochenkov** [comment]
>tokio::test adds #[test] so it would catch

Yes, `tokio::test` adds `#[test]`, but it happens later.
The first `#[test]` in
```rust
#[test]
#[tokio::test]
fn foo() {}
```
will see `#[tokio::test]` in unexpanded form.

**Comment by Centril** [comment]
Please use some other lint than `ill_formed_attribute_input` since that one is a C-future-compat lint moving towards a hard error (and the name doesn't seem appropriate either).

**Comment by petrochenkov** [comment]
`unused_attributes` would work too.

**Comment by gagan0723** [comment]
Ping from triage, @DoumanAsh any updates on this?

**Pull request closed without merging**